packages/twenty-server/src/engine/api/graphql/workspace-query-runner/workspace-pre-query-hook/workspace-query-hook.explorer.ts

```diff
@@ -0,0 +1,142 @@
+import { Injectable, Logger, OnModuleInit } from '@nestjs/common';
+import { DiscoveryService, ModuleRef, createContextId } from '@nestjs/core';
+import { Module } from '@nestjs/core/injector/module';
+import { Injector } from '@nestjs/core/injector/injector';
+
+import { WorkspaceQueryHookType } from 'src/engine/api/graphql/workspace-query-runner/workspace-pre-query-hook/interfaces/workspace-query-hook.type';
+import { WorkspaceQueryHookInstance } from 'src/engine/api/graphql/workspace-query-runner/workspace-pre-query-hook/interfaces/workspace-query-hook.interface';
+
+import { WorkspaceQueryHookMetadataAccessor } from 'src/engine/api/graphql/workspace-query-runner/workspace-pre-query-hook/workspace-query-hook-metadata.accessor';
+import { WorkspaceQueryHookStorage } from 'src/engine/api/graphql/workspace-query-runner/workspace-pre-query-hook/storage/workspace-query-hook.storage';
+import { WorkspaceQueryHookKey } from 'src/engine/api/graphql/workspace-query-runner/workspace-pre-query-hook/decorators/workspace-query-hook.decorator';
+
+@Injectable()
+export class WorkspaceQueryHookExplorer implements OnModuleInit {
+  private readonly logger = new Logger('WorkspaceQueryHookModule');
+  private readonly injector = new Injector();
+
+  constructor(
+    private readonly moduleRef: ModuleRef,
+    private readonly discoveryService: DiscoveryService,
+    private readonly metadataAccessor: WorkspaceQueryHookMetadataAccessor,
+    private readonly workspaceQueryHookStorage: WorkspaceQueryHookStorage,
+  ) {}
+
+  onModuleInit() {
+    this.explore();
+  }
+
+  async explore() {
+    console.log('EXPLORE !');
+    const hooks = this.discoveryService
+      .getProviders()
+      .filter((wrapper) =>
+        this.metadataAccessor.isWorkspaceQueryHook(
+          !wrapper.metatype || wrapper.inject
+            ? wrapper.instance?.constructor
+            : wrapper.metatype,
+        ),
+      );
+
+    console.log('HOOKS: ', hooks.length);
+
+    for (const hook of hooks) {
+      const { instance, metatype } = hook;
+
+      const { key, type } =
+        this.metadataAccessor.getWorkspaceQueryHookMetadata(
+          instance.constructor || metatype,
+        ) ?? {};
+
+      if (!key || !type) {
+        this.logger.error(
+          `PreHook ${hook.name} is missing key or type metadata`,
+        );
+        continue;
+      }
+
+      if (!hook.host) {
+        this.logger.error(`PreHook ${hook.name} is missing host metadata`);
+
+        continue;
+      }
+
+      this.registerWorkspaceQueryHook(
+        key,
+        type,
+        instance,
+        hook.host,
+        !hook.isDependencyTreeStatic(),
+      );
+    }
+  }
+
+  async handleHook(
+    payload: Parameters<WorkspaceQueryHookInstance['execute']>,
+    instance: object,
+    host: Module,
+    isRequestScoped: boolean,
+  ) {
+    const methodName = 'execute';
+
+    if (isRequestScoped) {
```

Comment:
note: I'm not sure we have any use case where we would have a pre-hook not scoped since hooks are applied over the flexible schema resolvers 🤔

Reply:
Hmm yes, I need to double test that to verify that it's never going through the else